csrc/expr_simplifier.cpp

```diff
@@ -2490,6 +2490,104 @@ Val* factorizeGcd(Val* value, const Context& context) {
   return value;
 }
 
+Val* cancelTermsInPredicate(Val* value, const Context& context) {
+  auto bop = dynamic_cast<BinaryOp*>(value->definition());
+  if (!bop) {
+    return value;
+  }
+  auto op = bop->getBinaryOpType();
+  if (op != BinaryOpType::Eq && op != BinaryOpType::GE &&
+      op != BinaryOpType::GT && op != BinaryOpType::LE &&
+      op != BinaryOpType::LT && op != BinaryOpType::NE) {
+    return value;
+  }
+
+  auto get_terms = [](Val* operand) -> std::vector<Val*> {
+    if (auto add_op = dynamic_cast<BinaryOp*>(operand->definition());
+        add_op != nullptr && add_op->getBinaryOpType() == BinaryOpType::Add) {
+      return add_op->inputs();
+    } else if (auto sub_op = dynamic_cast<BinaryOp*>(operand->definition());
+               sub_op != nullptr &&
+               sub_op->getBinaryOpType() == BinaryOpType::Sub) {
+      auto inputs = sub_op->inputs();
+      inputs.at(1) = IrBuilder::negExpr(inputs.at(1));
+      return inputs;
+    } else if (auto flattened = toFlattenedAdd(operand->definition())) {
+      return flattened->inputs();
+    } else if (auto flattened = assoc_comm::flatten(operand);
+               flattened->isA<FOp>() &&
+               flattened->as<FOp>()->getOpType() == BinaryOpType::Add) {
+      return flattened->as<FOp>()->inputs();
+    } else {
+      return {operand};
+    }
```

Comment:
We only need the `if (auto flattened = toFlattenedAdd(operand->definition()))` case and else case here. In the context of pass, associative and commutative ops are guaranteed to be flattened.

Reply:
I must be something wrong then since that doesn't seem to be the case. I inserted an assertion at each of the unnecessary paths, which indeed was triggered.

```
diff --git a/csrc/expr_simplifier.cpp b/csrc/expr_simplifier.cpp
index 5bcb1c1f..3190e833 100644
--- a/csrc/expr_simplifier.cpp
+++ b/csrc/expr_simplifier.cpp
@@ -2531,10 +2531,12 @@ Val* cancelTermsInPredicate(Val* value, const Context& context) {
   auto get_terms = [](Val* operand) -> std::vector<Val*> {
     if (auto add_op = dynamic_cast<BinaryOp*>(operand->definition());
         add_op != nullptr && add_op->getBinaryOpType() == BinaryOpType::Add) {
+      TORCH_INTERNAL_ASSERT(false);
       return add_op->inputs();
     } else if (auto sub_op = dynamic_cast<BinaryOp*>(operand->definition());
                sub_op != nullptr &&
                sub_op->getBinaryOpType() == BinaryOpType::Sub) {
+      TORCH_INTERNAL_ASSERT(false);
       auto inputs = sub_op->inputs();
       inputs.at(1) = IrBuilder::negExpr(inputs.at(1));
       return inputs;
@@ -2543,6 +2545,7 @@ Val* cancelTermsInPredicate(Val* value, const Context& context) {
     } else if (auto flattened = assoc_comm::flatten(operand);
                flattened->isA<FOp>() &&
                flattened->as<FOp>()->getOpType() == BinaryOpType::Add) {
+      TORCH_INTERNAL_ASSERT(false);
       return flattened->as<FOp>()->inputs();
     } else {
       return {operand};
```

```
 ./bin/nvfuser_tests  --gtest_filter="ExprSimplifierTest.Compare" 2>&1 |tee log
Only one CUDA device detected. Disabling MultiCUDA tests
Note: Google Test filter = ExprSimplifierTest.Compare-*_MultiCUDA
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ExprSimplifierTest
[ RUN      ] ExprSimplifierTest.Compare
unknown file: Failure
C++ exception with description "false INTERNAL ASSERT FAILED at "/home/nmaruyama/nvdl/nvfuser/debug2/csrc/expr_simplifier.cpp":2539, please report a bug to PyTorch.
Exception raised from operator() at /home/nmaruyama/nvdl/nvfuser/debug2/csrc/expr_simplifier.cpp:2539 (most recent call first):
frame #0: <unknown function> + 0xf6290 (0x7fd3c4e0c290 in /raid/nmaruyama/debug2/pytorch/build/lib/libc10.so)
```

Anything I'm doing wrong?

Reply:
Which case is failing? I guess it is the `auto flattened = assoc_comm::flatten(operand)` case? If so I think it is because if you already have a `FOp`, then flatten will return the input as is, and this path will be taken?

Reply:
This should be fixed in https://github.com/NVIDIA/Fuser/pull/705.

Reply:
Thanks. It is fixed by the PR.